### PR TITLE
Refactor PauliString.negated into complex PauliString.coefficient

### PR DESCRIPTION
**Pull request by Strilanc**
- Work around sympy not folding -1.0 * symbol into -symbol
- Don't quote circuit symbol equations like "-a" and "a*2". Just bracket them.
- Drop 'negate' method (redundant with unary negation)
- Leave some TODOs work in contrib
- Fix PauliStringGateOperation avoiding negating sympy expressions
- Add support for multiplying pauli strings by complex coefficients

### CODE REVIEW

File: cirq/contrib/paulistring/clifford_optimize.py [hunk 1]
```diff
@@ -100,7 +100,8 @@ def try_merge_clifford(cliff_op: ops.GateOperation, start_i: int) -> bool:
 
             qubit, pauli = next(iter(merge_op.pauli_string.items()))
             quarter_turns = round(merge_op.half_turns * 2)
-            quarter_turns *= (1, -1)[merge_op.pauli_string.negated]
+            assert merge_op.pauli_string.coefficient in [1, -1]
```
Comment (viathor) [inline cirq/contrib/paulistring/clifford_optimize.py]: Alternatively, you could keep the `negated` method and make it check the coefficient value, raise if coefficient's imaginary part is non-zero, else return its real part.

File: cirq/contrib/paulistring/pauli_string_phasor.py [hunk 2]
```diff
@@ -79,30 +80,26 @@ def can_merge_with(self, op: 'PauliStringPhasor') -> bool:
...
-                yield ops.X(any_qubit)
-            yield ops.Z(any_qubit)**self.half_turns
-            if self.pauli_string.negated:
-                yield ops.X(any_qubit)
-        else:
-            half_turns = self.half_turns * (-1 if self.pauli_string.negated
-                                               else 1)
-            yield ops.Z(any_qubit) ** half_turns
+        sign = self.pauli_string.coefficient.real
+        yield ops.Z(any_qubit)**protocols.mul(self.half_turns, sign)
```
Comment (viathor) [inline cirq/contrib/paulistring/pauli_string_phasor.py]: Let `self.pauli_string.coefficient` == exp(1j * theta). When theta is zero we want to emit `Z**1`. When theta is pi we want to emit `Z**-1`.

Isn't it true that in general, we'd want to emit Z**(theta/pi) here? If so, we should be able to remove the assertion above and handle all coefficient values here.

File: cirq/contrib/paulistring/pauli_string_raw_types.py [hunk 3]
```diff
@@ -63,10 +63,7 @@ def _pauli_string_diagram_info(self,
         qubits = self.qubits if args.known_qubits is None else args.known_qubits
         syms = tuple('[{}]'.format(self.pauli_string[qubit])
                      for qubit in qubits)
-        if exponent_absorbs_sign and self.pauli_string.negated:
-            if isinstance(exponent, float):
-                exponent = -exponent
-            else:
-                exponent = '-{!s}'.format(exponent)
+        if exponent_absorbs_sign and self.pauli_string.coefficient == -1:
```
Comment (viathor) [inline cirq/contrib/paulistring/pauli_string_raw_types.py]: This looks suspicious. What if coefficient is nearly -1, e.g. `exp(1j * pi * 99/100)`?

At minimum, we should raise. Better, find a way to handle all phases.

File: cirq/protocols/mul.py [hunk 4]
```diff
@@ -54,10 +54,14 @@ def mul(lhs: Any, rhs: Any, default: Any = RaiseTypeErrorIfNotProvided) -> Any:
...
-    if lhs == 1.0 and is_parameterized(rhs):
+    if lhs == 1 and is_parameterized(rhs):
         result = rhs
-    if rhs == 1.0 and is_parameterized(lhs):
+    if rhs == 1 and is_parameterized(lhs):
         result = lhs
+    if lhs == -1 and is_parameterized(rhs):
+        result = -rhs
+    if rhs == -1 and is_parameterized(lhs):
+        result = -lhs
```
[the remark this answers is not in the document]
Comment (Strilanc) [reply cirq/protocols/mul.py]: Sympy already handles that correctly.

File: cirq/contrib/paulistring/pauli_string_phasor.py (hunk 2 above)
Comment (Strilanc) [reply cirq/contrib/paulistring/pauli_string_phasor.py]: There is a continuous extension, but it requires generalizing pauli string phasor. Thus the TODO earlier in the method.

File: cirq/contrib/paulistring/pauli_string_raw_types.py (hunk 3 above)
Comment (Strilanc) [reply cirq/contrib/paulistring/pauli_string_raw_types.py]: Added a TODO.

This refactoring is somewhat large and I want to tackle it in pieces.

File: cirq/contrib/paulistring/clifford_optimize.py (hunk 1 above)
Comment (Strilanc) [reply cirq/contrib/paulistring/clifford_optimize.py]: Added a TODO.

File: cirq/ops/pauli_string_test.py [hunk 5]
```diff
@@ -39,45 +41,45 @@ def test_eq_ne_hash():
...
+               cirq.PauliString({q0: cirq.Y}, -1))
+    assert not cirq.PauliString({q0: cirq.X}, +1).equal_up_to_sign(
+               cirq.PauliString({q0: cirq.Y}, -1))
+
+    assert not cirq.PauliString({q0: cirq.X}, +1).equal_up_to_sign(
+               cirq.PauliString({}, +1))
+    assert not cirq.PauliString({q0: cirq.X}, -1).equal_up_to_sign(
+               cirq.PauliString({}, -1))
+    assert not cirq.PauliString({q0: cirq.X}, +1).equal_up_to_sign(
+               cirq.PauliString({}, -1))
```
[the remark this answers is not in the document]
Comment (Strilanc) [reply cirq/ops/pauli_string_test.py]: Done.

File: cirq/ops/pauli_string.py [hunk 6]
```diff
@@ -223,19 +242,24 @@ def _pass_pauli_interaction_gate_over(pauli_map: Dict[raw_types.Qid,
...
                                           after_to_before: bool = False
                                           ) -> bool:
-        def merge_and_kickback(qubit, pauli_left, pauli_right, inv):
+        def merge_and_kickback(qubit,
+                               pauli_left: Optional[pauli_gates.Pauli],
+                               pauli_right: Optional[pauli_gates.Pauli],
+                               inv: bool) -> int:
             if pauli_left is None or pauli_right is None:
-                pauli_map[qubit] = pauli_left or pauli_right
+                pauli_map[qubit] = cast(pauli_gates.Pauli,
```
[the remark this answers is not in the document]
Comment (Strilanc) [reply cirq/ops/pauli_string.py]: Added an assert.

File: cirq/ops/pauli_string.py [hunk 7]
```diff
@@ -171,17 +189,18 @@ def pass_operations_over(self,
...
                 # what op is.
                 continue
-            inv ^= PauliString._pass_operation_over(pauli_map,
-                                                    op,
-                                                    after_to_before)
-        return PauliString(pauli_map, inv)
+            should_negate ^= PauliString._pass_operation_over(pauli_map,
+                                                              op,
+                                                              after_to_before)
+        coef = self._coef * (-1 if should_negate else +1)
```
[the remark this answers is not in the document]
Comment (Strilanc) [reply cirq/ops/pauli_string.py]: Done

File: cirq/ops/pauli_string.py [hunk 8]
```diff
@@ -30,18 +30,25 @@
...
     def __init__(self,
                  qubit_pauli_map: Mapping[raw_types.Qid, Pauli],
-                 negated: bool = False) -> None:
+                 coefficient: Union[int, float, complex] = 1) -> None:
+        assert coefficient is not False
+        assert coefficient is not True
+        assert isinstance(coefficient, (int, float, complex))
         self._qubit_pauli_map = dict(qubit_pauli_map)
-        self.negated = negated
+        self._coef = complex(coefficient)
```
[the remark this answers is not in the document]
Comment (Strilanc) [reply cirq/ops/pauli_string.py]: Done.

File: cirq/ops/pauli_string.py [hunk 9]
```diff
@@ -30,18 +30,25 @@
 class PauliString(raw_types.Operation):
     def __init__(self,
                  qubit_pauli_map: Mapping[raw_types.Qid, Pauli],
-                 negated: bool = False) -> None:
+                 coefficient: Union[int, float, complex] = 1) -> None:
+        assert coefficient is not False
+        assert coefficient is not True
+        assert isinstance(coefficient, (int, float, complex))
```
[the remark this answers is not in the document]
Comment (Strilanc) [reply cirq/ops/pauli_string.py]: This was debug code I temporarily added to catch things passing in a boolean for 'negated`. Removed.

File: cirq/ops/pauli_string_test.py [hunk 10]
```diff
@@ -39,45 +41,45 @@ def test_eq_ne_hash():
...
-        eq.add_equality_group(cirq.PauliString({q: pauli}, True))
+        eq.add_equality_group(cirq.PauliString({q: pauli}, +1))
+        eq.add_equality_group(cirq.PauliString({q: pauli}, -1))
     for q, p0, p1 in itertools.product((q0, q1), (cirq.X, cirq.Y, cirq.Z),
                                        (cirq.X, cirq.Y, cirq.Z)):
-        eq.add_equality_group(cirq.PauliString({q: p0, q2: p1}, False))
+        eq.add_equality_group(cirq.PauliString({q: p0, q2: p1}, +1))
 
 
 def test_equal_up_to_sign():
```
[the remark this answers is not in the document]
Comment (Strilanc) [reply cirq/ops/pauli_string_test.py]: Done.

**Pull request merged**